Tiling drag, take 3

orestisfl opened the pull request:
Fixes #2643.
This is a continuation of #2653.

~~I have changed the behaviour for the inner region from #2653. The inner region now behaves like `move container to mark` and supports no direction. This reduces the changes to core `con.c` functions and is simpler. It was also enough for the support of stacked & tabbed containers without any extra work.~~

The code needs some polishing here and there but I hope that the core functionality works as intended with zero crashes. I have opened this PR to get some first feedback and hopefully get some people to run the code on their machines.

## Current status
### TODOs
(removed completed items from #2653):

- [X] Support stacks and tabs
- [X] Drag onto the root window
- [x] ~~Initiate tiling drag from titlebar~~ This would break resizing
- [x] Emit move event properly on tiling drop.❓I need some help here, I don't know what the original intention was?
- [x] ~~Refactor insert_con_into() into con.c~~ Why?
- [x] ~~Configure color of drop indicator❓ (even though using the focused color seems like a reasonable default, do we want to add the extra configuration?)~~
- [X] Bug: i3 crashes if a window closes while the dragging is indicating to drop it onto that window.
- [x] Accidentally mod + clicking a window can be annoying. Maybe only show the indicator after we have dragged the container outside of its original position.
- [x] Add a way to disable the whole thing
- [x] Only focus `con` if it had focus. To change this, `route_click` should also be modified since focus changes there.
- [x] I think the indicator sizes need some work. The outer indicators are big in small windows and small in big windows.
- [x] I am not sure if fb4fa338b6b741733c93599c95b1b2df61b602b7 is needed. Investigate?
- [x] Moving an unfocused container from another output shouldn't make it focused #3200
- [x] Resize from title bar w/ mouse broke I think
- [x] For consistency: should dragging unfocused floating containers also not make them focused?
- [X] ~~Rearrange tabbed / stacked by dropping on the title?~~ => out of scope
- [x] ~~dragging & swapping ?~~ -- can be done in future PR
- [x] Documentation: indicators + disable option.
- [x] Squash commits better to remove unneeded changes from con.c, potentially cluttering `git blame`.
- [x] Switch to meson
- [x] Update release notes

### Tests
- [x] #3193 
- [x] dragging onto a bar
- [x] dragging onto an empty workspace
- [x] move events
- [x] dragging onto a container that closes before the drag is complete
- [x] config option
- [x] ~~dragging onto stacked / tabbed containers~~ => out of scope
- [ ] more complicated layouts with a combination of orientations etc
- [x] moving an unfocused container doesn't make it focused
- [x] moving an unfocused container from another output doesn't make it focused
- [x] moving fullscreen containers
- [ ] focus order
- [x] update filename

Airblader commented:
> Emit move event properly on tiling drop.

I think @acrisci just meant that when you drop-move a window, we need to emit the regular `window::move` event. Perhaps this wasn't the case in the other PR(s) yet, perhaps it just hadn't been tested yet; I'm not sure.

> Configure color of drop indicator

Let's hold off on that for now. I think the focused color is a reasonable choice. When the need arises we can still make it configurable.

>  I am not sure if fb4fa33 is needed. Investigate?

Perhaps @MForster can add some insight here. :-) I haven't looked into it yet.

> Squash commits better to remove unneeded changes from con.c, potentially cluttering git blame (?).

Yes, let's please squash; but we can wait with that until we merge. To retain attributions, the commit message should just mention the other authors having worked on this previously.

> Add a way to disable the whole thing?

That's a very good point. Yes, I think we should have this option.

> I have opened this PR to get some first feedback and hopefully get some people to run the code on their machines.

Perhaps leave a comment at the existing two PRs so the people subscribed there get a notification that this has been picked up upon again? You can also drop an email to the mailing list and / or call for testers on reddit.

orestisf1993 commented:
> > I have opened this PR to get some first feedback and hopefully get some people to run the code on their machines.
> 
> You can also drop an email to the mailing list and / or call for testers on reddit.

We can probably do this when this is ready to merge (or after it is merged in a separate branch for limited time? dunno). I predict that writing excessive tests for this will take some time.

For now, it would be great if you and/or @stapelberg can run this.

MForster commented:
> > I am not sure if fb4fa33 is needed. Investigate?
> 
> Perhaps @MForster can add some insight here. :-) I haven't looked into it yet.

I'm sure that it it was necessary when I implemented it :-). There may be simpler solutions.

orestisf1993 commented:
I've been trying to write the tests. Maybe @stapelberg can help with my problem:

```perl
use i3test i3_config => <<EOT;
# i3 config file (v4)
font -misc-fixed-medium-r-normal--13-120-75-75-C-70-iso10646-1

focus_follows_mouse no
floating_modifier Mod1

# 2 side by side outputs
fake-outputs 1000x500+0+0,1000x500+1000+0
EOT
use i3test::XTEST;

sub start_drag {
    my ($pos_x, $pos_y) = @_;
    # $x->root->warp_pointer($pos_x, $pos_y);
    # sync_with_i3;
    xtest_key_press(64);        # Alt_L
	xtest_button_press(1, $pos_x, $pos_y);
    xtest_sync_with_i3;
}

sub end_drag {
    my ($pos_x, $pos_y) = @_;
    $x->root->warp_pointer($pos_x, $pos_y);
    # sync_with_i3;
    # xtest_sync_with_i3;
	xtest_button_release(1, $pos_x, $pos_y);
    xtest_key_release(64);      # Alt_L
    xtest_sync_with_i3;
    # sync_with_i3;
}

my $ws2 = fresh_workspace(output => 1);
my $ws1 = fresh_workspace(output => 0);
open_window;
my $A = get_focused($ws1);

start_drag(50, 50);
end_drag(1050, 50);

is(get_focused($ws2), $A, 'A moved to the right workspace');

done_testing;
```

I have tried every sync permutation for the `end_drag` subroutine. I think the sync after `warp_pointer` is needed but including it will sometimes make the test stuck, waiting for the sync forever.

orestisf1993 commented:
Hmm, I think it gets stuck because i3 is inside the `tiling_drag` function and can't reply to the sync message. Don't know how I can work around that.

orestisf1993 commented:
Possible feature: special keybind that when used with tiling windows it turns them to floating.

From https://www.reddit.com/r/i3wm/comments/7xrmxu/move_nonfloating_window_using_the_mouse/

orestisf1993 commented:
@stapelberg, since you are fairly active lately, do you have any input regarding the problem I described above?

Thanks.

stapelberg commented:
Sorry for the late reply.

I think your assessment is correct: we don’t correctly handle the sync event from within the drag_pointer event loop. It should be an easy addition to do so, though :). Let me know if you can’t get that to work, and thanks for being diligent regarding the tests — I think this will be our first test case performing any kind of dragging.

orestisf1993 commented:
They generally seem to be handled but they get stuck sometimes. The test sometimes hangs for 10-20 seconds and then continues. This is the gdb log.

```
(gdb) p (xcb_generic_event_t*) xcb_poll_for_event(conn)
$1 = (xcb_generic_event_t *) 0x604000016350
(gdb) p $1->response_type & 0x7F
$2 = 28
(gdb) p (xcb_generic_event_t*) xcb_poll_for_event(conn)
$3 = (xcb_generic_event_t *) 0x604000016390
(gdb) p $3->response_type & 0x7F
$4 = 33
(gdb) bt
#0  0x00007f801996df90 in epoll_pwait () at /usr/lib/libc.so.6
#1  0x00007f801a1a4e95 in  () at /usr/lib/libev.so.4
#2  0x00007f801a1a7455 in ev_run () at /usr/lib/libev.so.4
#3  0x0000555fa2b54f39 in drag_pointer (con=0x614000002440, event=0x604000013a10, confine_to=0, border=BORDER_TOP, cursor=8, callback=0x555fa2ba1f34 <drag_callback>, extra=0x7ffc393e7f60) at ../../i3/src/floating.c:848
#4  0x0000555fa2ba3043 in tiling_drag (con=0x614000002440, event=0x604000013a10) at ../../i3/src/tiling_drag.c:187
#5  0x0000555fa2b03d3f in route_click (con=0x614000002440, event=0x604000013a10, mod_pressed=true, dest=CLICK_INSIDE) at ../../i3/src/click.c:261
#6  0x0000555fa2b046cc in handle_button_press (event=0x604000013a10) at ../../i3/src/click.c:364
#7  0x0000555fa2b61707 in handle_event (type=4, event=0x604000013a10) at ../../i3/src/handlers.c:1551
#8  0x0000555fa2b71f9c in xcb_prepare_cb (loop=0x7f801a3ada40, w=0x60300000d6f0, revents=16384) at ../../i3/src/main.c:133
#9  0x00007f801a1a44e3 in ev_invoke_pending () at /usr/lib/libev.so.4
#10 0x00007f801a1a7fe1 in ev_run () at /usr/lib/libev.so.4
#11 0x0000555fa2b71d45 in ev_loop (loop=0x7f801a3ada40, flags=0) at /usr/include/ev.h:835
#12 0x0000555fa2b7a95c in main (argc=9, argv=0x7ffc393e9848) at ../../i3/src/main.c:964
```

It seems like it's possible to have leftover X events but ev_run doesn't get interrupted?

I'll continue tomorrow.

orestisf1993 commented:
Ok, I lied

```
diff --git a/src/floating.c b/src/floating.c
index e958153d..ce769195 100644
--- a/src/floating.c
+++ b/src/floating.c
@@ -692,6 +692,7 @@ static void xcb_drag_prepare_cb(EV_P_ ev_prepare *w, int revents) {
     xcb_motion_notify_event_t *last_motion_notify = NULL;
     xcb_generic_event_t *event;
 
+    while (1) {
         while ((event = xcb_poll_for_event(conn)) != NULL) {
             if (event->response_type == 0) {
                 xcb_generic_error_t *error = (xcb_generic_error_t *)event;
@@ -768,9 +769,11 @@ static void xcb_drag_prepare_cb(EV_P_ ev_prepare *w, int revents) {
                 dragloop->extra);
         }
         free(last_motion_notify);
+        last_motion_notify = NULL;
 
         xcb_flush(conn);
     }
+}
 
 /*
  * This function grabs your pointer and keyboard and lets you drag stuff around
```

this quick hack solves the problem.

This is a race condition and should probably be fixed properly.

stapelberg commented:
Just checking — your git working directory does contain commit 0d8b6714e39af81cbd6f4fbad500872a715dea24, yes? If not, that would be my suspicion.

orestisf1993 commented:
Yes, I had rebased with the latest next.

orestisf1993 commented:
I wonder if we are going against libev's documentation here:

> You must not call ev_run (or similar functions that enter the current event loop) or ev_loop_fork from either ev_prepare or ev_check watchers. Other loops than the current one are fine, however. The rationale behind this is that you do not need to check for recursion in those watchers, i.e. the sequence will always be ev_prepare, blocking, ev_check so if you have one watcher of each kind they will always be called in pairs bracketing the blocking call.

stapelberg commented:
Fair point. If you can find a cleaner way to implement this, pull requests are very welcome :)

orestisf1993 commented:
Tbh, I don't really understand the rationale behind the quoted text, I don't think that the current implementation causes any problems as far as I can tell. I'll see if I can come up with a cleaner solution though.

For this PR though my problem is that I still don't completely understand why the i3sync event gets lost. I have tried some hacks that improved the situation a bit (eg an additional ev_idle or ev_check makes the 1-minute hang more rare) but I am still searching for a complete solution.

Since you are significantly more experienced with libev any ideas for the possible solution/cause might help.

stapelberg commented:
Can you give me the steps required to reproduce the issue please?

orestisf1993 commented:
I've pushed a test that should exhibit this behaviour.

stapelberg commented:
I looked into this a little bit, and I’m fairly convinced that the issue is that we’re draining the events in https://github.com/i3/i3/blob/ddc80ab3054711269c7875de2bad916ba9c93ef1/src/floating.c#L695, but don’t handle the case where new events become available as part of the dragloop callback in https://github.com/i3/i3/blob/ddc80ab3054711269c7875de2bad916ba9c93ef1/src/floating.c#L763: to process replies, xcb needs to read from the X11 connection, which might mean reading new events.

Indeed, adding a quick and dirty “drain:” label in front of the while loop and jumping to it before the final `xcb_flush(conn)` call in `xcb_drag_prepare_cb` fixes the hang for me.

Note that this doesn’t eliminate all causes of flakiness for 295-tiling-drag.t for me. One other issue is that you’re calling `xtest_sync_with_i3` after calling `$x->root->warp_pointer`, which uses the wrong X11 connection: the XTEST code has a separate X11 connection. So, make that `sync_with_i3`. Even with that fix, the test is still flaky: sometimes i3 does not seem to receive any pointer update events. Not entirely sure why.

I hope this helps you make some progress on this. Let me know if you need further help.

orestisf1993 commented:
Thanks! I had done something similar with your `drain` but I didn't know about i3_sync. I combined:
1. the `drain` label
2. the sync fix
3. ev_run once -> ev_loop

I ran it hundreds of times with a loop and it never hanged.

orestisf1993 commented:
I think we should make the functionality a bit more complex for this to be more useful. I don't like the (original) idea of always introducing splits though because it could get annoying quick.

I am thinking something like:
1. Inner area works like `move to mark` - what we are doing right now.
2. Outer area should always put the dragged container in the same parent with the target. If the orientation matches, don't introduce a new split (current behaviour). If it doesn't put them in a new split.
3. Maybe use another area that covers the parent container, drawn over more than one windows.
Eg with `V[A B] C` when I want to move C to the left of V the drop indicator should cover the left side of A and B.

We could also use the `swap` command for $mod + Shift. Sway does something like that: continuously moves the window you are dragging. So, if the layout is `A B C D` dragging C onto A will make the new layout `C A B D`. Instead of that we could only swap with the target, so dragging C onto A would make the layout `C B A D`. Or have a way to do both, this is not too complex.

@Airblader since you are interested in this feature can you test it locally for a bit to provide your thoughts?

Airblader commented:
I would love to, but I'm just a bit short on time. :-( This issue originally generated a lot of interest in the community. Perhaps we should advertise it again for some discussion?

orestisf1993 commented:
How it looks right now: https://gfycat.com/UnevenFocusedBaleenwhale

sirsurthur commented:
Gif looks cool. Great work !

stapelberg commented:
Looks really good indeed!

Airblader commented:
Looks really cool indeed! I saw some questions in your original post:

> For consistency: should dragging unfocused floating containers also not make them focused?

I think it's reasonable and expected that dragging a window also focuses it.

> Rearrange tabbed / stacked by dropping on the title?

Sure, why not. But we can also do it in a second PR.

orestisf1993 commented:
hmm, I am getting flakyness with the tests again :/

Airblader commented:
@orestisf1993 What's the status here?

orestisf1993 commented:
I am still getting flakyness in the tests and I haven't found the cause for it yet. Maybe we can change the testing approach? Eg, assume the mouse will work correctly and call the tiling functions through an IPC command, eg `tiling_drag 50 50 100 100` (original x y & target x y).

Airblader commented:
That'd be fine with me.

orestisf1993 commented:
I removed the 'stale' label since I plan to finish this. Would be a shame to close this when the main code is almost ready.

tksfz commented:
I was among the people who initially requested this feature. I'm really looking forward to this! Go go go!

orestisf1993 commented:
You can always apply the changes locally, I don't know of any bugs currently. Feedback would be very welcome as well.

The reason this has slowed down is that while testing I spot some issues elsewhere that have to be fixed and then I have to continue working on the tests etc.

livanh commented:
I really want to see this in i3. Take your time, but don't give up! I'll test it locally and report any weird thing I might notice.

Airblader commented:
@orestisf1993 A year later, do we still expect work on this to continue? :-)

orestisf1993 commented:
I wanted to continue with https://github.com/i3/i3/issues/3503 first so that I can write more reliable tests regarding tree layouts but we can postpone more complicated tests for later and aim to merge this as it is.

I'll restart contributing to i3 soon, so I am pretty certain that I'll finish work on this.

Airblader commented:
I'm fine with postponing to improve the PR with other changes first. If you say this continues, that's good enough for me. :-)

odinsdream commented:
Oh I'm super excited to have found that this work is in progress, and really hope it continues. I just moved from SwayWM to i3 due to changing away from Wayland, and Sway has this drag feature that I am already sorely missing. It makes it so quick and easy to reorder your tiles and rearrange how they work. Definitely would love to see this moved forward, thank you all!

clapbr commented:
I am quadriplegic and use a trackball+pointer so I can only use 2-key combinations or 1 key+mouse combination like this PR implements. This feature made me possible to use i3 confortably, thank you very much.

I've been using it as a daily driver for the last week moving windows around exclusively with this feature and never encountered any issue.

Airblader commented:
@clapbr Thank you for testing the PR and reporting back how it is working for you, thus is very valuable feedback!

Ursescu commented:
Hi guys I love this feature (tested on personal machine). What is the current status for this PR, any chances that will get this in the next release ?

orestisfl commented:
Status pretty much the same. Planning to get back to it in late August or
in September. Hopefully will finish before the next release.

Thanks for testing it and if you have any feedback they are more than
welcome.

On Wed, Aug 7, 2019, 17:04 Ursescu Ionut <notifications@github.com> wrote:

> Hi guys I love this feature (tested on personal machine). What is the
> current status for this PR, any chances that will get this in the next
> release ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/i3/i3/pull/3085?email_source=notifications&email_token=ABMCZPSRHYIIDGSWQVB5AADQDLI6ZA5CNFSM4EH3YDJ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3YQNFQ#issuecomment-519112342>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABMCZPTGY5HUY7WUHBI3533QDLI6ZANCNFSM4EH3YDJQ>
> .
>

clapbr commented:
Is the small conflict with current git easily fixable?

orestisfl committed:
Allow dragging tiled windows with the mouse.

Fixes #2643

Inner drop region behaves like move to mark.
The outer region is close to the edge (currently 30px from the edge).
This will place the container as a sibling in the given direction within
the parent container. If the move direction goes against the orientation
of the parent container, tree_move() is called.

Contributors:
Co-authored-by: Orestis Floros <orestisflo@gmail.com>
See #3085
- Inner drop region behaves like move to mark
- Handle workspaces
- Fix crash when target closes
- Initiate tiling drag from titlebar
- Hide indicator until container is dragged outside of original position
- Calculate outer_threshold using percentages instead of fixed pixel
values
- Emit 'move' event properly
- Don't focus previously unfocused containers
- Use tree_split() on different orientation
- Fix redundant split containers
- DT_PARENT
- Readability & optimizations
- Limit parent threshold by render_deco_height()
- Tests
- Fullscreen container handling
- Initiate drag from title bar
- Fix issue of EnterNotify events still triggering after drag_callback
  is called
- Include decorations for drop target calculation

Co-authored-by: Michael Forster <email@michael-forster.de>
See #2178
- Original implementation of tiling drag + indicator window
> A container can be dragged by the title bar to one of the four sides
> of another container. That container will then be split either
> horizontally or vertically.

Co-authored-by: Tony Crisci <tony@dubstepdish.com>
See #2653
- Original implementation of outer/inner drop region indicator:
> There are two drop regions per direction.
>
> The inner region is closer to the center of the window. Dropping on
> this region will split the target container and put the container
> within the split at the given direction beside the target container.
>
> The outer region is close to the edge (currently 30px from the edge).
> This will place the container as a sibling in the given direction within
> the parent container.
>
> Dropping into the outer region moves the con beside the target. If the
> move direction goes against the orientation of the parent container, the
> con moves out of the row.
- Fix crash: Ignore containers without a managed window (eg i3bar)

orestisfl commented:
@clapbr I fixed the conflict

orestisfl commented:
@snikz

> Ive tested this for a week now, and it would be a really good addition if you dropped a window in the middle of another window they changed position. It somewhat works in some situations, but should be a 1 to 1 change of position and size.
> Other than that it works really well.

Does the swap command do what you want? To try it select a target. run `i3-msg mark a`, then select the other container and run the command `i3-msg swap with mark a`. 
If yes, I plan to also support swap for tiling drop by holding the Shift modifier. Not in this PR though but it should be trivial enough once this gets merged.

Diegovsky commented:
Is this pull request still working? I wanted to give it a drive and return some feedback

botiapa commented:
Any progress on this PR? Would love to test it if needed.

ghost commented:
@orestisfl Are you still working on this?

orestisfl commented:
I haven't had time to work on this but I regularly think about finishing this.

orestisfl commented:
@Airblader I rebased. Not sure if there is any point in adding more tests here since this uses i3 internals and is hard to test because of mouse movements. Main thing left is documentation.

Airblader commented on src/tiling_drag.c:
```diff
@@ -0,0 +1,328 @@
+ */
+static Con *find_drop_target(uint32_t x, uint32_t y) {
+    Con *con;
+    /*TODO: also decoration*/
```
Open TODO?

Airblader commented on src/tiling_drag.c:
```diff
@@ -0,0 +1,328 @@
+     * position. */
+    const uint32_t min_rect_dimension = min(rect.width, rect.height);
+    const uint32_t sibling_threshold = max(1, (uint32_t)(0.3 * min_rect_dimension));
+    const uint32_t parent_threshold = max(1, min(render_deco_height() + 5, (uint32_t)(0.15 * min_rect_dimension)));
```
What's the meaning of this magic number 5 in here? The 0.15 and 0.3 factors are also a bit magical. Can we give them names?

Airblader commented on src/tiling_drag.c:
```diff
@@ -0,0 +1,328 @@
+    const uint32_t d_right = rect.x + rect.width - new_x;
+    const uint32_t d_bottom = rect.y + rect.height - new_y;
+    const uint32_t d_min = min(min(d_left, d_right), min(d_top, d_bottom));
+    /* And move the container towards that direction. */
```
Could we maybe move this into a static function to declutter this long function a bit?

orestisfl replied:
I have no idea how to cleanly separate this :)

Airblader replied:
~L155–L174 depend only on `rect`, `new_x`, and `new_y` to determine `direction`. So couldn't we make this a function of three arguments returning the direction?~

Ah, I only now saw that `d_min` is used afterwards still.

Airblader commented on src/tiling_drag.c:
```diff
@@ -0,0 +1,328 @@
+
+create_indicator:
+    if (draw_window) {
+        if (*(params->indicator) == 0) {
```
NULL?

orestisfl replied:
No, it should be `xcb_window_t`

Airblader commented on include/configuration.h:
```diff
@@ -110,6 +110,12 @@ struct Config {
+     * mouse button to move tiling containers. The user can turn this
+     * functionality off.
+     */
+    bool disable_tiling_drag;
```
I wonder whether we really need to make this configurable. This currently does nothing, and it works only on a decoration that i3 provides. Am I missing something or why would this need to be a configuration flag?

orestisfl replied:
Affects `click.c` in two points. More backwards compatible for users that want to propagate $mod+ click to the windows.

Airblader replied:
If it breaks existing behavior, the feature needs to be off by default, not the other way around.

orestisfl replied:
I think most users will benefit from this being on by default. The slightly incompatible behavior will not be that common but it's good to have a way out.

orestisfl replied:
I think most users will benefit from this being on by default. The slightly incompatible behavior will not be that common but it's good to have a way out.

Airblader replied:
I don't disagree on it being useful, but not breaking existing workflows is one of our most important project goals, so much so that it is documented on the homepage of the website. IMO we cannot simultaneously say this is important enough to add configuration, but not important enough to make the default compatible.

orestisfl replied:
I would rather remove the configuration than hide one of the [most popular](https://github.com/i3/i3/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) requests behind an option that a large amount of people won't see.

> but not breaking existing workflows is one of our most important project goals

Sure, but it's not really the first time we make minor changes to behavior, especially in `click.c`-related functionality. Only very edge cases would be affected.

Anyway, I removed the option and we'll see if there are going to be people that request it back.

Airblader replied:
> I would rather remove the configuration than hide one of the most popular requests behind an option that a large amount of people won't see.

One way would've been to make it an opt-in, but add it to the default config. 

> More backwards compatible for users that want to propagate $mod+ click to the windows.

Just to be clear, this is only for clicks on the decoration, not the client window itself, right?

orestisfl replied:
Either clicks on the title bar or mod+clicks on the window.

Default config also sounds fine for me

Airblader replied:
Hm, mod+click sounds like it has some potential to interfere with things – that said, we have mod+right click, and that doesn't.

Maybe we go without configuration for now since it is easier to add than remove configuration, but we should keep this in mind before a release.

orestisfl replied:
$mod+click is already used for moving floating containers

Airblader replied:
Fair point. But then going without configuration for now seems even more appropriate.

Airblader reviewed (commented):
@orestisfl I just gave this a quick scan. It looks good for the most part, couple minor comments.

Most importantly, I'd like to test this first (which I haven't) to assess the look&feel a bit. Functionality-wise I doubt I'll manage to test it excessively, so given the interest in this on this PR if anyone else here would be able to run this for a while, feedback on how you feel about it and any issues you encountered would be very welcome.

adam4235 commented:
I'd be happy to test this and give feedback.  Does that mean I need to build the code?  I tried downloading the branch and building it according to the instructions [here](https://i3wm.org/docs/hacking-howto.html) but I got some errors.  Where should I ask for help with the build?  (I've never built i3 before.)

orestisfl commented:
@adam4235 Thanks, that would be helpful. What errors did you get? Make sure you also have installed the dependencies that you will need: https://github.com/Airblader/i3/wiki/Building-from-source

Airblader commented:
If your distribution has some git/vcs package for i3, a good alternative would also be to use that but just exchange the git URL used for building.

Ursescu commented:
Hi, I would also like to test this out. (I actually did when this first came out, but my distro package manager removed my hand crafted i3-package). I would like to know if this is a conflict free merge on top of i3-gaps (what I am currently using).

Airblader commented:
@Ursescu Seems that way. I've pushed the branch `tiling-drag` on the i3-gaps repository which is the current i3-gaps HEAD with this branch merged in. (Edit: to clarify, I only did this for convenience for people who want to try this out but use i3-gaps, I don't intend to maintain that branch; we either merge it here or nowhere)

adam4235 commented:
> @adam4235 Thanks, that would be helpful. What errors did you get? Make sure you also have installed the dependencies that you will need: https://github.com/Airblader/i3/wiki/Building-from-source

Thanks, it was a dependency issue and I got it built now.  I'd suggest that the Hacking How To should be updated with the dependencies or a link to them.

adam4235 commented:
I've been trying it out for a bit.  First, some background on the way I use i3: I set "workspace_layout tabbed" in my config file, and mostly just have a bunch of windows in that tab group, with only one filling the screen at a time, and switch between them, like a bunch of maximized windows in a floating window manager.  If I want to see 2 at a time horizontally, I move a window to the right of all the other windows, or move it below to arrange vertically.  I rarely / never use h/v split containers, because my use cases aren't complicated enough to need them.

Here are some initial impressions of the tiling drag:

-I wasn't expecting to have to hold the Windows ($mod) key to do the dragging.  My naive expectation was to be able to click and drag the title bar of a window in order to move that window, i.e. drag it to the right edge of the screen to split horizontally and to the bottom of the screen to split vertically.  In fact, at first I didn't think I was using the modified i3 executable because nothing happened when I tried to do that.  (Ideally, it would also make sense to drag one title bar onto another title bar to switch their places in a tabbed layout, but that seems like it should be a separate change.)

Since people from floating window managers are used to moving windows by clicking and dragging the titlebar, it would make sense for clicking and dragging the titlebar to also move windows in a tiling window manager, for a different definition of "move".

-Once I arranged windows horizontally or vertically by moving a window to the right or below, I had some difficulty putting it back.  After trial and error it seems that in normal circumstances the way to do that is to drag one window to the middle of the other one.  However, in some cases I ended up creating a V or H split by accident, and once I did I found it difficult to get rid of since I never use them.  I was expecting dragging one to the center of the other to do so, as it does when I've arranged them horizontally / vertically by a move command, but dragging to the center did nothing once I had (accidentally) created a split container.  After reading the design at [this](https://github.com/i3/i3/pull/2653#issuecomment-274371597) comment, I figured out that I had created the H/V split by moving to the wrong region, and to get it back I had to drag to the outer region to do a move.

-Relatedly, if I have many windows open and want to split one horizontally, I can't do that with a tiling drag, or for that matter, the keyboard, from within my tabbed layout, unless the one I want to split is the rightmost (or leftmost) in the tabbed group.  I have to move the window all the way to the right (or left) and then do one more move to cause the split.  I normally do this by pushing a keyboard shortcut repeatedly to move the window all the way to to the right until it splits (maybe this is a suboptimal and weird way of doing this and is unrelated to this bug?).  With the mouse, this would take many drag and drops and isn't feasible.  But, if it were the case that dragging the titlebar of a window over other titlebars in a tabbed layout caused them to switch places, then I could drag the window all the way to the right and let go to do the horizontal split.

Hopefully that provides some feedback from the perspective of someone who uses i3 in a naive / simplistic way.  Basically, my simplistic use case would be to drag a window to the right to move it to the right beside my other windows, or drag it down to move it below the other windows.  Then when I'm done looking at 2 windows at once, drag it back over the other windows to put it back.  It would be great if there was an option to just have 4 regions to drag to, instead of 8, so that this simple use case wouldn't accidentally introduce weird effects.

clapbr commented:
> -I wasn't expecting to have to hold the Windows ($mod) key to do the dragging. My naive expectation was to be able to click and drag the title bar of a window in order to move that window, i.e. drag it to the right edge of the screen to split horizontally and to the bottom of the screen to split vertically. In fact, at first I didn't think I was using the modified i3 executable because nothing happened when I tried to do that. (Ideally, it would also make sense to drag one title bar onto another title bar to switch their places in a tabbed layout, but that seems like it should be a separate change.)
> 
> Since people from floating window managers are used to moving windows by clicking and dragging the titlebar, it would make sense for clicking and dragging the titlebar to also move windows in a tiling window manager, for a different definition of "move".

That's how sway does it FWIW. Works great for me (quadriplegic with a trackball)

Airblader commented:
Thanks @adam4235 for the feedback! That's very valuable.

> I wasn't expecting to have to hold the Windows ($mod) key to do the dragging. 

I agree with that expectation. It's also consistent with dragging floating windows. Unless there's a reason I'm not aware of (@orestisfl ?) we should change that.

orestisfl commented:
@Airblader I see the TODOs have a checked "Initiate tiling drag from titlebar". It probably broke somewhere in the process. I'll see what's wrong.

orestisfl commented:
Oh I think there is an important reason why tiling drag cannot be initiated from the titlebar: dragging the titlebar resizes inside vertical containers

adam4235 commented:
> Oh I think there is an important reason why tiling drag cannot be initiated from the titlebar: dragging the titlebar resizes inside vertical containers

Hmm, I note that there is currently a discrepancy in resizing windows with the mouse.  When resizing windows that are arranged horizontally, I have to grab a tiny border between them (it's so tiny that I sometimes miss and have to try again).  But when resizing windows that are arranged vertically, I can grab the title bar, which is much larger.  (I think some people use i3-gaps to get around this discrepancy.)

I also note that in floating window managers, and for that matter floating windows in i3, resizing involves grabbing a border.

Would it be fair to say that the reason clicking and dragging the title bar currently causes a resize is because, until now, there was nothing else for it to do?

Airblader commented:
I think I agree with what I assume @adam4235 is getting at; for vertical splits, we could only let the border do the resize, but dragging the titlebar itself could do the tiling drag. I think that's a reasonable change since horizontal splits behave that way and thus this is more or less making it more consistent and using the fruits of that for the new feature. WDYT @orestisfl ?

Ursescu commented:
@Airblader Thanks for the i3 gaps branch, worked like a charm.

For now I have one observation, I would like that when I want to move a window on one edge of the screen, I can drag the mouse to the end of the monitor. I think this is a problem with gaps because when I put the cursor on the gap at the edge, the tilling window drag white guideline will disappear.

https://user-images.githubusercontent.com/15110622/104123751-a1fd7500-5355-11eb-955b-665a47123802.mp4

Airblader commented:
Yeah, unfortunately I think i3-gaps will require some additional changes for this after it is merged into i3. This is a good thing to be aware of. I don't know I'll have the time to address this myself, so either @orestisfl wants to fix this in i3-gaps afterwards as well or I might have to stub out this feature until someone can contribute a fix. We'll see.

orestisfl commented:
> I think I agree with what I assume @adam4235 is getting at; for vertical splits, we could only let the border do the resize, but dragging the titlebar itself could do the tiling drag. I think that's a reasonable change since horizontal splits behave that way and thus this is more or less making it more consistent and using the fruits of that for the new feature. WDYT @orestisfl ?

But there is no border on the container with the title. You would have to focus the window on top first and then drag its border.

ra-om commented:
> > I think I agree with what I assume @adam4235 is getting at; for vertical splits, we could only let the border do the resize, but dragging the titlebar itself could do the tiling drag. I think that's a reasonable change since horizontal splits behave that way and thus this is more or less making it more consistent and using the fruits of that for the new feature. WDYT @orestisfl ?
> 
> But there is no border on the container with the title. You would have to focus the window on top first and then drag its border.

With focus_follows_mouse disabled, if you click right on the border on the unfocused window you can still resize on the first click. I don't see a problem.

ra-om commented:
I've made a script that provides most of this functionality, including tab/stack support. The only significant lack I think are the visual indicators in real-time.

![mvwm](https://user-images.githubusercontent.com/12862942/114310606-44dea000-9ab9-11eb-8d42-12ba418388fb.png)

roughly the zones where the window is going to end up, each margin is 40% of the window's dimension.

The script is designed to right-click then drag the window, but it's easy to change to any other button.

https://github.com/ra-om/scripts
Bind the file mouse-id to i3's config and point it to MoveWindowsWithMouse, I was too lazy to reimplement the little bash script in python.

I lack the time and knowledge to help implement this PR, but maybe this'll be enough for the impatient.

orestisfl commented:
@Airblader ready :)

orestisfl commented:
@Airblader ok, ready (again)

Airblader commented on docs/userguide:
```diff
@@ -196,6 +196,53 @@ provided by the i3 https://github.com/i3/i3/blob/next/etc/config.keycodes[defaul
+=== Moving tiling containers with the mouse
+
+Since i3 4.21, it's possible to drag tiling containers using the mouse. The
+drag can be initiating either by dragging the window's titlebar or by pressing
```
```suggestion
drag can be initiated either by dragging the window's titlebar or by pressing
```

Airblader commented on src/tiling_drag.c:
```diff
@@ -0,0 +1,357 @@
+    const uint32_t sibling_indicator_size = max(
+        /* 30% of the container (minus the parent indicator) is used to drop
+         * the dragged container as a sibling to the target: */
+        (uint32_t)(0.3 * min_rect_dimension),
```
Maybe we can move these constants to the top level and document them there; I suspect some people will want to change them, and while I don't think we should do so via configuration, making them constants not only aids readability but also "hackability".

orestisfl replied:
Done

stapelberg commented:
Got a chance to try this out. Here are a few notes:

* The highlighting of the drag destination is good, but I wonder if we could add short explanatory text, like “add to this container” or “create new split container”? It’s not 100% intuitive to me which target does what (I think I figured out most targets, but…), and I bet the same is true for many users, too :) (Could be done in a follow-up change, too.)
* I managed to crash i3 with certain drags. Not sure if the PR should already be crash-free, so let me know if you can’t reproduce and I can grab more details.

orestisfl commented:
Michael, thank you for your points.

1. That's a nice idea but I would prefer to have it as a follow-up.
2. Yes, this should be crash-free maybe I introduced a bug with recent changes. Would love to have some reproduction instructions.

stapelberg commented:
Sounds good. I’ll try again to reproduce the crashes and let you know. Good work on this PR!

orestisfl commented:
Improved IMO behavior regarding tabbed / stacked parents when using the outer indicator. Now, `con` will move next to the tabbed target.

old: https://user-images.githubusercontent.com/5778622/141200767-eb3721a3-9c85-4956-a48f-34b7fe4cf510.mp4

new: https://user-images.githubusercontent.com/5778622/141200124-9f0359d1-c61d-4e2f-9465-ecec609c653b.mp4

stapelberg commented:
> Sounds good. I’ll try again to reproduce the crashes and let you know. Good work on this PR!

Here’s the backtrace:
https://gist.github.com/stapelberg/e81415dca5287ebbbe3de326e09bafb7

Here’s the log:
https://logs.i3wm.org/logs/6283859616333824.bz2

I just attached a debugger and started dragging windows around without much thought. The debugger stopped soon enough. (Not sure if this helps to reproduce, just wanted to mention it.)

Let me know if the above isn’t sufficient and I can try going at this in a more deliberate way.

orestisfl commented:
Thanks Michael. Now it's trivial to reproduce :) Surprised this didn't happen to me.

Pushed a fix.

stapelberg commented:
> Pushed a fix.

Thanks, I can no longer trigger a crash now :)

victordiaz commented:
any advancement on this? I think it might be incredibly useful for mouse-based workflows!

orestisfl commented:
Waiting for a review AFAIK

Airblader commented:
I will try to review this soon (understandably this is a bit more work here). Of course it would be extremely helpful for others to run this locally to capture any potential crashes.

stapelberg commented on src/tiling_drag.c:
```diff
@@ -0,0 +1,379 @@
+
+/*
+ * The callback that is executed on every mouse move while dragging. On each
+ * invocation we determin the drop target and the direction in which to insert
```
nit: determin → determine

stapelberg commented on src/tiling_drag.c:
```diff
@@ -0,0 +1,379 @@
+    /* Use the base decoration height and add a few pixels. This makes the
+     * outter indicator generally thin but at least thick enough to cover
+     * container titles */
+    const double parent_indicator_max_size = render_deco_height() + 5;
```
Here and below, shouldn’t we use `logical_px(5)` for this code to work equally well on a hi-dpi display?

stapelberg commented on src/tiling_drag.c:
```diff
@@ -0,0 +1,379 @@
+     * dragged container as a sibling to the target */
+    const double sibling_indicator_percent_of_rect = 0.3;
+    /* Use the base decoration height and add a few pixels. This makes the
+     * outter indicator generally thin but at least thick enough to cover
```
nit: outter → outer

stapelberg commented on src/tiling_drag.c:
```diff
@@ -0,0 +1,379 @@
+    direction_t direction;
+    drop_type_t drop_type;
+    xcb_window_t indicator = 0;
+    const struct callback_params params = {&indicator, &target, &direction, &drop_type};
```
Is it really required to use pointer fields within `struct callback_params`?

Could you instead turn this into `struct callback_params params = {
.target = NULL,
.direction = 0,
.drop_type = 0,
.indicator = 0,
}` and just use e.g. `params->indicator = …` within `drag_pointer()`?

orestisfl replied:
`DRAGGING_CB` is defined as

```
#define DRAGGING_CB(name)                                                      \
    static void name(Con *con, Rect *old_rect, uint32_t new_x, uint32_t new_y, \
                     const xcb_button_press_event_t *event, const void *extra)
```

so if we modify the struct we will discard the `const` qualifier. So our only option without modifying the option signature is to directly write to these set addresses in memory.

stapelberg replied:
Thanks for the explanation. I think we can remove the `const` qualifier, which will result in better code overall — if the params need to be mutable in this case, we might as well reflect that in the signature.

stapelberg commented on src/tiling_drag.c:
```diff
@@ -0,0 +1,379 @@
+
+            ipc_send_window_event("move", con);
+            break;
+        case DT_PARENT:;
```
nit: extra `;` at the end of line?

orestisfl replied:
This is a C-ism: https://stackoverflow.com/questions/8384388/variable-declaration-after-goto-label

For whatever reason my gcc allows it but clang complains:

```
clang -Ii3.p -I. -I.. -I../include -I/usr/include/startup-notification-1.0 -I/usr/include/yajl -I/usr/include/cairo -I/usr/include/lzo -I/usr/include/libpng16 -I/usr/include/freetype2 -I/usr/include/harfbuzz -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -I/usr/include/sysprof-4 -I/usr/include/pixman-1 -I/usr/include/pango-1.0 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/fribidi -fcolor-diagnostics -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=c11 -O0 -g -Wunused-value -D_GNU_SOURCE -pthread -MD -MQ i3.p/src_tiling_drag.c.o -MF i3.p/src_tiling_drag.c.o.d -o i3.p/src_tiling_drag.c.o -c ../src/tiling_drag.c
../src/tiling_drag.c:324:13: error: expected expression
            const bool parent_tabbed_or_stacked = (target->parent->layout == L_TABBED || target->parent->layout == L_STACKED);
            ^
../src/tiling_drag.c:328:18: error: use of undeclared identifier 'parent_tabbed_or_stacked'
                 parent_tabbed_or_stacked ? "tabbed/stacked " : "",
                 ^
../src/tiling_drag.c:330:17: error: use of undeclared identifier 'parent_tabbed_or_stacked'
            if (parent_tabbed_or_stacked) {
```

stapelberg replied:
TIL :) 

Thus far, I have usually created a new scope to be able to declare variables after a label:

```c
case DT_PARENT: {
  const bool …
}
```

Can you see if that works here? I would prefer a new scope over the subtle `;`. If a new scope doesn’t work, at least add a comment explaining the `;`

stapelberg reviewed (commented):
Did another review pass, but I think we can merge this after the comments are addressed. I’d rather have this merged and see bug reports that we can then fix before the next release, instead of dragging this feature out any further :)

Airblader commented:
@orestisfl Would you be able to merge & test this in i3-gaps as well to make sure everything works (no idea whether this is even any effort).

dechimp commented:
I've been daily-driving i3 on @orestisfl's branch for over a month without noticing any issues. I have multiple monitors as well. Excited to see this merge.

opalmay commented:
I've been running this with i3-gaps, it mostly works, but dragging into the middle section doesn't consistently act like moving containers to marks. Maybe I'm misusing it or misunderstanding the feature somehow.

orestisfl commented:
Can you give examples please?

On Sat, 30 Apr 2022, 11:35 opalmay, ***@***.***> wrote:

> I've been running this with i3-gaps, it mostly works, but dragging into
> the middle section doesn't consistently act like moving containers to
> marks. Maybe I'm misusing it or misunderstanding the feature somehow.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/i3/i3/pull/3085#issuecomment-1113957808>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABMCZPT32DX5CIEV6P6GZWDVHT5EVANCNFSM4EH3YDJQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

phwelo commented:
> Can you give examples please?
> […](#)
> On Sat, 30 Apr 2022, 11:35 opalmay, ***@***.***> wrote: I've been running this with i3-gaps, it mostly works, but dragging into the middle section doesn't consistently act like moving containers to marks. Maybe I'm misusing it or misunderstanding the feature somehow. — Reply to this email directly, view it on GitHub <[#3085 (comment)](https://github.com/i3/i3/pull/3085#issuecomment-1113957808)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABMCZPT32DX5CIEV6P6GZWDVHT5EVANCNFSM4EH3YDJQ> . You are receiving this because you were mentioned.Message ID: ***@***.***>

I am testing this on `gaps-next`.  The middle-box inconsistencies are certainly there.  I tried to capture them in a video here https://youtu.be/s9-4OS-GgvA

As far as I can tell so far, it seems like dragging to any other position besides center works great.  unfortunately it's hard to predict the behavior of the center mark right now.

TheSunCat commented:
I'm running this fork and love how it works! However it seems resizing tiled windows seems to have a chance of crashing. I can reproduce a crash by opening two windows side-by-side and resizing them repeatedly for a few seconds. This didn't happen on stock i3 built from here.
i3 gives me an option to save a backtrace, but I'm not certain where this is saved. If I find it, I'll attach it in an edit to this comment.

stapelberg commented:
Please see https://i3wm.org/docs/debugging.html#_on_crashes_obtaining_a_backtrace for details regarding the backtrace. Thanks for providing one, that would be super helpful!

TheSunCat commented:
Well this is awkward... the backtrace file just contains the text `No stack.`
I have GDB installed, and I checked that the i3 executable was not stripped.

dechimp commented:
Kind of curious what causes the crashes for you.  I've been daily-driving
orestisfl/i3 for 3+ months and havent noticed anything abnormal.

On Tue, May 24, 2022 at 2:28 PM RealTheSunCat ***@***.***>
wrote:

> Well this is awkward... the backtrace file just contains the text No
> stack.
> I have GDB installed, and I checked that the i3 executable was not
> stripped.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/i3/i3/pull/3085#issuecomment-1136399930>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFWJVZ65HX2ZKSRD3H34UDVLU3XDANCNFSM4EH3YDJQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

TheSunCat commented:
Me, too! It's weirdly consistent, as well. I am able to reproduce it by just opening a new window and resizing it a couple of times. After restarting i3 in place (pressing r), it does not crash anymore.

I've attached a video of the crash.
https://user-images.githubusercontent.com/44881120/170129578-9cdb706a-8df1-439e-98c7-65c14627ad44.mp4

stapelberg commented:
> Well this is awkward... the backtrace file just contains the text `No stack.`
> I have GDB installed, and I checked that the i3 executable was not stripped.

Weird. When you switch to a text console (or SSH into your machine) and attach gdb to the i3 process, does `backtrace full` result in stack traces? 

> I am able to reproduce it by just opening a new window and resizing it a couple of times. After restarting i3 in place (pressing r), it does not crash anymore.

Can you add a new user account to your computer (so that all configs are back to their defaults) and reproduce it with a fresh i3 config, too? If yes, can you provide a debug log of your i3 session up until the crash? That might contain enough clues to track this down. See https://i3wm.org/docs/debugging.html for details on logging.

orestisfl commented:
> Well this is awkward... the backtrace file just contains the text `No stack.` I have GDB installed, and I checked that the i3 executable was not stripped.

You probably need `echo 0 | sudo tee /proc/sys/kernel/yama/ptrace_scope`

TheSunCat commented:
> You probably need `echo 0 | sudo tee /proc/sys/kernel/yama/ptrace_scope`

Ah-ha! That allowed me to get a backtrace! I've attached it here: 
[i3-backtrace.717.0.txt](https://github.com/i3/i3/files/8782140/i3-backtrace.717.0.txt)

> Can you add a new user account to your computer (so that all configs are back to their defaults) and reproduce it with a fresh i3 config, too?

I created a new user with an empty home, and was unable to reproduce it... so it's probably something about my setup that causes this crash. Hrm. I'm willing to spend the time to help debug this, but I don't remember exactly how I set up what I'm currently running, so I'm not sure I can easily recreate a MWE.

TheSunCat commented:
> You probably need `echo 0 | sudo tee /proc/sys/kernel/yama/ptrace_scope`

Yup, that gave me a real backtrace, thanks! Would be nice to add this to the i3 docs. I've attached it [here](https://github.com/i3/i3/files/8789426/i3-backtrace.722.1.txt). Looks like the `resize_graphical_handler` is causing the crash here.

> Can you add a new user account to your computer (so that all configs are back to their defaults) and reproduce it with a fresh i3 config, too?

Huh! I added an account with an empty home, and this does not occur. I guess this means something is wrong with my setup specifically. I'm willing to help debug this, although recreating my setup step-by-step on this new account would be difficult as I have a lot of parts that I added since first installing it.

orestisfl commented:
> > Can you give examples please?
> > […](#)
> > On Sat, 30 Apr 2022, 11:35 opalmay, _**@**_._**> wrote: I've been running this with i3-gaps, it mostly works, but dragging into the middle section doesn't consistently act like moving containers to marks. Maybe I'm misusing it or misunderstanding the feature somehow. — Reply to this email directly, view it on GitHub <[#3085 (comment)](https://github.com/i3/i3/pull/3085#issuecomment-1113957808)>, or unsubscribe https://github.com/notifications/unsubscribe-auth/ABMCZPT32DX5CIEV6P6GZWDVHT5EVANCNFSM4EH3YDJQ . You are receiving this because you were mentioned.Message ID: _**@**_.**_>
> 
> I am testing this on `gaps-next`. The middle-box inconsistencies are certainly there. I tried to capture them in a video here https://youtu.be/s9-4OS-GgvA
> 
> As far as I can tell so far, it seems like dragging to any other position besides center works great. unfortunately it's hard to predict the behavior of the center mark right now.

From a quick look in the video, behavior seems to be correct. With i3, you always need to be aware of the tree layout, even when dragging with the mouse. Dragging onto the center makes the dragged container a sibling of the target.

orestisfl commented:
> > You probably need `echo 0 | sudo tee /proc/sys/kernel/yama/ptrace_scope`
> 
> Yup, that gave me a real backtrace, thanks! Would be nice to add this to the i3 docs. I've attached it [here](https://github.com/i3/i3/files/8789426/i3-backtrace.722.1.txt). Looks like the `resize_graphical_handler` is causing the crash here.
> 
> > Can you add a new user account to your computer (so that all configs are back to their defaults) and reproduce it with a fresh i3 config, too?
> 
> Huh! I added an account with an empty home, and this does not occur. I guess this means something is wrong with my setup specifically. I'm willing to help debug this, although recreating my setup step-by-step on this new account would be difficult as I have a lot of parts that I added since first installing it.

I am having trouble to reproduce the issue and by looking at the stack trace, I can't see how this is tiling-drag-specific.

@stapelberg @Airblader WDYT about merging this to get more testers?

stapelberg commented:
Sounds good to me. I already mentioned last time that once comments are addressed, we can merge this :)

Airblader commented:
No concerns merging it into i3. I do have concerns with i3-gaps since there seems to be odd behavior, but nothing I can do about that, I guess.

stapelberg commented on testcases/t/316-drag-container.t:
```diff
@@ -0,0 +1,322 @@
+
+my ($width, $height) = (1000, 500);
+
+# use i3test i3_config => <<"EOT";
```
remove left-over

stapelberg reviewed (commented):
Looks good overall. Just a few small finishing touches.

orestisfl commented:
Regarding the `const` qualifier (can't reply in-thread for some reason), the benefit of the old method is that it allows to specify both read-only and read-write arguments, otherwise it's not easy to tell just by looking at the struct.

I have made the change but I would still favor the old code 1. because of the point above 2. because it minimizes risk of a last-minute regression for this PR.

stapelberg commented:
> Regarding the `const` qualifier (can't reply in-thread for some reason), the benefit of the old method is that it allows to specify both read-only and read-write arguments, otherwise it's not easy to tell just by looking at the struct.
> 
> I have made the change but I would still favor the old code 1. because of the point above 2. because it minimizes risk of a last-minute regression for this PR.

That’s fair. We can leave the `const` qualifier in.

Can you update and rebase to pick up the CI fixes please?

orestisfl commented:
Done :)

stapelberg commented:
Merged! 🎉

Thanks for your work on this!

vincentbernat commented:
Thanks for the patch! I have just tested it (on top of 4.20.1) and I have found a case where it does not work:

```
(splitv)
 (splith)
  (splith)
    emacs
    (splitv)
      vbeterm
      vbeterm
```

If I drag the top vbeterm on top of the bottom vbeterm, the positions of the two are swapped as expected. But if I drag the bottom vbeterm on top of the top one, the two windows are not swapped. The indicator is correct (I get an orange area on the center of the target term), but when releasing the mouse button, it does nothing.

Edit: I have tested on top of i3-gaps, so maybe this does not happen with vanilla i3. I'll test on top of vanilla i3 a bit later.

orestisfl commented:
@vincentbernat Thanks for trying this already

The semantics of drag & drop are not to swap containers, but similar to `move to mark`. So, the behavior is correct.

Copy & paste from the docs in this PR:

```
Drop on container::
    This happens when the mouse is relatively near the center of a container.
    If the mouse is released, the result is exactly as if you had run the
    +move container to mark+ command. See <<move_to_mark>>.
```

vincentbernat commented:
Understood, thanks for the clarification!

MForster commented:
Thanks a lot for seeing this through to completion! I always felt bad about teasing the feature with an unfinished patch and then letting it drop. Seeing how much work this was I wouldn't have been able to to see it through.

So, thanks again for picking this up and finishing it!

autoteelar commented:
how do i disable it? 

i looked in documentation and it tells you pretty much everything but that

stapelberg commented:
Can you explain why you want to disable the feature, as opposed to just not using it? Is it getting in your way somehow?

orestisfl commented:
> how do i disable it?
> 
> i looked in documentation and it tells you pretty much everything but that

You can't. Please track updates on a possible setting for this here: https://github.com/i3/i3/issues/5155

autoteelar commented:
> Can you explain why you want to disable the feature, as opposed to just not using it? Is it getting in your way somehow?

yeah, well i usually click the  tabs to change  them, and now that this is added if you move your mouse within like 3px while clicking it drags them off. 

i also move my windows with keyboard

basically tldr, i don't plan on using it, and it gets in the way in a pretty big way in my specific use case. i can understand why it was added and why other people might want it and see it as a useful feature as well though,

and well, it was also at least intended to have a way to disable this feature if you scroll to the top, airblader agreed its a good idea. (quoted here)


    「Add a way to disable the whole thing?

That's a very good point. Yes, I think we should have this option.」

i guess that about sums it up.

ra-om commented:
Left click is maybe not the best choice. Would be nice to be able to change which mouse button triggers the drag.

orestisfl commented:
It's probably time to lock this thread to encourage proper structured discussion via new issues. This is a released feature now so future changes need to be treated as bugs/enhancements.

Thank you to everyone that contributed to this discussion over the years, feedback has been very valuable!
